### ISSUE TITLE
Error when Export decimal -  C#

### ISSUE THREAD

**Issue by git2013vb** [godotengine/godot#53071]
### Godot version

v3.4.beta5.mono.official [dd0ee4872]

### System information

Debian 11 - C#

### Issue description

I have this error in terminal/output tab

```
E 0:00:00.763   _get_member_export: Unknown exported member type: 'DesertEdge.Entities.Stat.AttackPerSecond'.
  <C++ Source>  modules/mono/csharp_script.cpp:2574 @ _get_member_export()

```

The line at the error refer is:

 ```
       [Export]
        [Size("4.2")]
        public decimal AttackPerSecond { get; set; }
```

where `[Size("4.2")]` is my custom attribute.
The custom attribute is necessary to define size's table field in database.

### Steps to reproduce

Just use decimal as type then export it I guess.
I'm not sure but I think is a regression.

### Minimal reproduction project

NA

**Comment by raulsntos** [comment]
I can reproduce in Godot 3.2 so I don't think this ever worked.

I guess the problem is in the [`managed_to_variant_type`](https://github.com/godotengine/godot/blob/c7f67daccdc389d5feb89775297cd521fb9af882/modules/mono/mono_gd/gd_mono_marshal.cpp#L40) method. The `p_type.type_encoding` for `System.Decimal` seems to be 17 (or `MonoTypeEnum::VALUE_TYPE`) which makes sense because `System.Decimal` is a struct. However, the method only checks for the Godot structs so I guess we'd have to also check if it's `System.Decimal` somehow and then return `Variant::FLOAT`? (but I'm not sure if it's that simple). Also, the methods [`mono_object_to_variant_impl`](https://github.com/godotengine/godot/blob/c7f67daccdc389d5feb89775297cd521fb9af882/modules/mono/mono_gd/gd_mono_marshal.cpp#L972) and [`variant_to_mono_object`](https://github.com/godotengine/godot/blob/c7f67daccdc389d5feb89775297cd521fb9af882/modules/mono/mono_gd/gd_mono_marshal.cpp#L457) would have to implement the conversion to be able to convert from `System.Decimal` to `Variant` and the other way around.

**Comment by git2013vb** [comment]
That make sense to me. I never be able to use a Struct in any rpc calls ( where they use Variant type). 
I'm not an expert but `decimal` is the same as `System.Decimal` in c#? its the same struct? Why is missing such important piece in godot? There is a reason or was a miss? :)) I'm just curious.

**Comment by raulsntos** [comment]
In C# there 3 [floating-point numeric types](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/floating-point-numeric-types):

- `System.Single` or `float` (32 bits)
- `System.Double` or `double` (64 bits)
- `System.Decimal` or `decimal` (128 bits)

As far as I know Variants store floating-point numbers as `double` (64 bits) so it's likely that it's not possible to convert from `System.Decimal` to `Variant` without losing precision.

Have you tried using `double` instead?

**Comment by gyroninja** [comment]
> How I can save a number like `2.06` in sqlite using c#? If I use float the value will be `2.0699999332428` . But if I use decimal the result is what I want. But decimal isn't supported :(

I just want to point out that sqlite does not support decimal either. The only floating point type it supports is double.

**Comment by git2013vb** [comment]
> > How I can save a number like `2.06` in sqlite using c#? If I use float the value will be `2.0699999332428` . But if I use decimal the result is what I want. But decimal isn't supported :(
> 
> I just want to point out that sqlite does not support decimal either. The only floating point type it supports is double.

My references about sqlite data types came from here: 
https://www.guru99.com/sqlite-data-types.html
@raulsntos  my current workaround is to set them as string the to do my conversions in code.. :)

**Comment by raulsntos** [comment]
I'm not familiar with SQLite, but from what I understand you are using a `NUMERIC` affinity which stores the value as an integer (32 bits) if it has no decimals or as a real (64 bits) if it has decimals.

If the value is stored as 64 bits in the database and your C# variable is of type `decimal` (128 bits) then it won't fit, which means some bits will have to be discarded and some precision can be lost.

GodotSharp is not avoiding the `System.Decimal` type, implementing the conversion is just not as simple as you might think. Also, support for a 128 bits floating-point number is not really common, C++ does not support it and since Godot is written in C++ the Variants store floating-point numbers in 64 bits floating-point numbers (the equivalent to C# `double`).

I think you might be misunderstanding because of the type names, different languages and libraries might use a different name for the same type or might store it differently. In the end, it all comes down to how many bits are used to store the value and how those bits translate into a base-10 number since values are stored in a computer in binary (base-2 number).

Having said that, I agree that `decimal` should be supported in GodotSharp, since it's one of the floating-point number representations in C#'s standard library. But, for your use case, it seems you can get away with simply using `double` instead.

**Comment by git2013vb** [comment]
> If the value is stored as 64 bits in the database and your C# variable is of type `decimal` (128 bits) then it won't fit, which means some bits will have to be discarded and some precision can be lost.

From Sqlite page 
 - INTEGER type affinity in SQLite can hold any assigned integer number (positive or negative) from 1 byte to maximum 8 bytes.
 - REAL numbers are the number with double floating points precision. SQLite stored real numbers as 8 bytes’ array
 - NUMERIC affinity is assigned for any other data type. (`bool`, `datetime`, and `decimal` :) )

So in my case , discard some bits is not wrong. but is the right things to do
$2.0699999332428  is wrong. Its much more that $2.06 . So I MUST discard  $0.0099999332428 . :) 

I can't use type `double` because in sqlite will be REAL too likewise `float`. 
Sqlite use `decimal` type synonym of `currency`
![image](https://user-images.githubusercontent.com/13235119/135027173-b538604b-7e2d-44a9-bf2a-11bd7dae45e4.png)

**Comment by raulsntos** [comment]
> NUMERIC affinity is assigned for any other data type. (bool, datetime, and decimal :) )

Yes, but like I said `decimal` can be ambiguous because what SQLite refers to with `decimal` is not the same type as C#'s `System.Decimal` and it is encoded differently.

From the SQLite documentation `NUMERIC` affinity stores the value as an integer (32 bits) if it has no decimals or as a real (64 bits) if it has decimals, which means that for the number `2.06` SQLite's `NUMERIC` is the same as C#'s `double`. That's why I recommended you to try `double` since the precision loss will be smaller, have you tried that yet?

> So in my case , discard some bits is not wrong. but is the right things to do
$2.0699999332428 is wrong. Its much more than $2.06 . So I MUST discard $0.0099999332428 . :)

Well, what you are describing is not discarding bits but discarding decimal units. The **IEEE 754** standard describes how to encode floating-point numbers.

A 32 bit floating-point number (`float`) is encoded as:
![image](https://user-images.githubusercontent.com/3903059/135095167-fc84eb15-8689-472b-b014-205a2b2e95d5.png)

The decimal number `2.06` is encoded in `float` as `0 10000000 00000111101011100001010`.
And encoded in `double` as `0 10000000000 0000011110101110000101000111101011100001010001111011`.

The `decimal` type in C# is different because it's more like four 32-bit integers (that's why it needs 128 bits). So `2.06` is encoded as `206 0 0 131072`.

> > implementing the conversion is just not as simple as you might think.
> Maybe is not so difficult - I'm not a c# guru of course :))

This doesn't have anything to do with C# since the conversion is actually implemented in C++ and it's about storing data represented by 128 bits in 64 bits (which will have some data loss no matter how you do it).

We could implement a `decimal` type in C++, of course, and the IEEE has a standard describing how to do so. However, since this is such an uncommon type I don't think that's going to happen. Also, the solution needs to work for every compiler since we don't want to force users to use a specific C++ compiler to use Godot.

---

Just to summarize:

- SQLite stores `2.06` as `double` (yes, that's how `NUMERIC` works).
- Exporting `System.Decimal` is difficult because Variant uses `double` and discarding bits reduces precision (it doesn't just remove decimal units).

**Comment by git2013vb** [comment]
Sure I will try it. For now I found and alternative  solution - I mentioned before. 

About the current issue I would suggest to "add" a layer to existing code available only who will use c# - so where the issue lie atm , I'm not aware if visual-script, native, or gdscript have same issue - in order to have the problem fixed (I guess the OS will be linux and windows?)
